### DISCUSSION TITLE
Question regarding browser engine detection

### DISCUSSION THREAD

**Discussion by chris124567** [fingerprintjs/fingerprintjs#641] [Q&A] [answered]
I've noticed there are some sources (`src/sources/chrome.ts` and `src/sources/error_ff.ts`) that do browser engine detection.  Why not just add these to the functions in `src/utils/browser.ts` and have a source called browser_engine that calls the functions in `src/utils/browser.ts` or something like that?

I also have another question regarding the functions in `src/utils/browser.ts`.  I was wondering how the values the  `countTruthy` result is compared to were decided.  For instance, in:

```
export function isChromium86OrNewer(): boolean {
  // Checked in Chrome 85 vs Chrome 86 both on desktop and Android
  const w = window

  return (
    countTruthy([
      !('MediaSettingsRange' in w),
      'RTCEncodedAudioFrame' in w,
      '' + w.Intl === '[object Intl]',
      '' + w.Reflect === '[object Reflect]',
    ]) >= 3
  )
}
```

How was 3 decided, and in:

```
export function isGecko(): boolean {
  const w = window

  // Based on research in September 2020
  return (
    countTruthy([
      'buildID' in navigator,
      'MozAppearance' in (document.documentElement?.style ?? {}),
      'MediaRecorderErrorEvent' in w,
      'mozInnerScreenX' in w,
      'CSSMozDocumentRule' in w,
      'CanvasCaptureMediaStream' in w,
    ]) >= 4
  )
}
```

how was 4 decided? Why does 1 or 2 not work - shouldn't any browser with window.mozInnerScreenX be called Firefox?

**Comment by Valve** [comment]
@chris124567 1 or 2 would work in a regular environment, but in a hostile environment it's not uncommon to see some attributes spoofed or overwritten, that's the reason we need more signals and more truthy values.
Regarding your other question @Finesse will answer later

**Comment by Finesse** [comment, marked as answer]
Hi, @chris124567. These are good questions.

> I've noticed there are some sources (`src/sources/chrome.ts` and `src/sources/error_ff.ts`) that do browser engine detection. Why not just add these to the functions in `src/utils/browser.ts` and have a source called browser_engine that calls the functions in `src/utils/browser.ts` or something like that?

You are right, these sources are excessive. We will remove them soon, in the next minor release. Check the [next-sources branch](https://github.com/fingerprintjs/fingerprintjs/tree/next-sources) to see the new list of entropy sources. We don't aim to make an entropy source that detects the browser engine because the current sources give enough entropy to tell one engine from another

> I also have another question regarding the functions in `src/utils/browser.ts`. I was wondering how the values the `countTruthy` result is compared to were decided. For instance: ... how was 4 decided? Why does 1 or 2 not work - shouldn't any browser with window.mozInnerScreenX be called Firefox?

At the moment, any single criterion used in the "is(EngineName)" functions should be able to be used to detect the engines. But a criterion can stop working correctly with a browser update, it can either stop detecting the engine or start detecting more than one engine. Also, the page JS code can affect the criteria results.

In order to protect the function from these unwanted effects, we've implemented a fuzzy logic with `countTruthy`. If a criterion inside a function returns a wrong result, the function result will stay correct. There is no strict rule for the numbers to compare `countTruthy` results with. I use an intuitive rule: most of the criteria must return `true` in order for a function to return `true`.
